### PR TITLE
refactor sampling layer setup

### PR DESCRIPTION
**Pull request by irexyc**
## Motivation

setup sampling layer after forward to increase parallelism

test with llama-2-7b-chat and 1000 num_prompts

before 
```
--------------------------------------------------
concurrency: 128
elapsed_time: 79.527s

number of prompt tokens: 248339
number of completion tokens: 240582
token throughput (completion token): 3025.156 token/s
token throughput (prompt + completion token): 6147.852 token/s
RPS (request per second): 12.574 req/s
RPM (request per minute): 754.460 req/min
--------------------------------------------------
```

after 
```
--------------------------------------------------
concurrency: 128
elapsed_time: 78.886s

number of prompt tokens: 248339
number of completion tokens: 240582
token throughput (completion token): 3049.727 token/s
token throughput (prompt + completion token): 6197.786 token/s
RPS (request per second): 12.676 req/s
RPM (request per minute): 760.587 req/min
--------------------------------------------------
```

### CODE REVIEW

File: src/turbomind/models/llama/LlamaBatch.cc [hunk 1]
```diff
@@ -1131,6 +1134,7 @@ void LlamaBatch<T>::InitializeSampling(const GenerationState& g)
         }
     }
     outputs_ = std::move(outputs);
+    sync_check_cuda_error();
```
Comment (zhyncs) [inline src/turbomind/models/llama/LlamaBatch.cc:1137]: Why should we `cudaDeviceSynchronize` here?

**Review by zhyncs — COMMENTED** [review]
6 unit tests have failed and may need to be fixed.

```
[  FAILED  ] 6 tests, listed below:
[  FAILED  ] SamplingDecodeTest/0.TopP, where TypeParam = float
[  FAILED  ] SamplingDecodeTest/0.BatchTopP, where TypeParam = float
[  FAILED  ] SamplingDecodeTest/0.InvalidArgsZeroTopP, where TypeParam = float
[  FAILED  ] SamplingDecodeTest/0.InvalidArgsBatchTopPContainZero, where TypeParam = float
[  FAILED  ] SamplingDecodeTest/0.InvalidArgsTopKBatchTopPContainZero, where TypeParam = float
[  FAILED  ] SamplingDecodeTest/0.LocalBatchBatchTopP, where TypeParam = float

 6 FAILED TESTS
```

File: src/turbomind/models/llama/LlamaBatch.cc (hunk 1 above)
Comment (irexyc) [reply src/turbomind/models/llama/LlamaBatch.cc:1137]: It will sync only when TM_DEBUG_LEVEL=DEBUG, I add here to better calculate the time

File: src/turbomind/layers/sampling_layers/TopPSamplingLayer.cu [hunk 2]
```diff
@@ -342,16 +273,6 @@ void TopPSamplingLayer<T>::runSampling(TensorMap* output_tensors, TensorMap* inp
         skip_decode_buf_ + ite * local_batch_size);
     sync_check_cuda_error();
 
-    invokeComputeToppDecay(
```
Comment (zhyncs) [inline src/turbomind/layers/sampling_layers/TopPSamplingLayer.cu:345]: Since the call to this kernel is no longer needed, we can consider deleting the corresponding implementation as well.
https://github.com/InternLM/lmdeploy/blob/5059f37c1db6f607c0723932f8f90834ae6b7ba2/src/turbomind/kernels/sampling_topp_kernels.h#L153
https://github.com/InternLM/lmdeploy/blob/5059f37c1db6f607c0723932f8f90834ae6b7ba2/src/turbomind/kernels/sampling_topp_kernels.cu#L1581

File: src/turbomind/layers/sampling_layers/TopPSamplingLayer.cu [hunk 3]
```diff
@@ -28,47 +28,16 @@
...
-                                             int             top_ps_size,
-                                             bool*           skip_decode,
-                                             float*          initial_top_p_buf,
-                                             float*          top_p_decay_buf,
-                                             const float*    top_p_decay,
-                                             float*          top_p_min_buf,
-                                             const float*    top_p_min,
-                                             int32_t*        top_p_reset_ids_buf,
-                                             const uint32_t* top_p_reset_ids)
+void set_topp_runtime_args(int    batch_size,
```
Comment (zhyncs) [inline src/turbomind/layers/sampling_layers/TopPSamplingLayer.cu:31]: nit: Before this PR, the kernel implemented using CUDA has been changed to a CPU function. The file extension can be changed from `.cu` to `.cc`. And the corresponding `CMakeLists.txt` also needs to be modified.

File: src/turbomind/layers/sampling_layers/TopPSamplingLayer.h [hunk 4]
```diff
@@ -30,13 +31,12 @@ class TopPSamplingLayer: public BaseSamplingLayer<T> {
...
     void freeBuffer() override;
 
-    uint*    runtime_top_k_buf_ = nullptr;
-    float*   runtime_top_p_buf_ = nullptr;
-    float    runtime_max_top_p_;
-    float*   initial_top_p_buf_   = nullptr;
-    float*   top_p_decay_buf_     = nullptr;
-    float*   top_p_min_buf_       = nullptr;
-    int32_t* top_p_reset_ids_buf_ = nullptr;
+    std::vector<uint>  h_runtime_top_k;
```
Comment (zhyncs) [inline src/turbomind/layers/sampling_layers/TopPSamplingLayer.h:34]: nit: The naming style is better kept consistent with the existing one, and use underscores for member variables at the end.

File: src/turbomind/layers/sampling_layers/TopPSamplingLayer.cu [hunk 5]
```diff
@@ -198,60 +137,52 @@ void TopPSamplingLayer<T>::setup(const size_t batch_size, const size_t beam_widt
...
     const Tensor runtime_top_p = runtime_args->isExist("runtime_top_p") ? runtime_args->at("runtime_top_p") : Tensor();
     const size_t runtime_top_p_size = runtime_top_p.size();
-    if (runtime_top_p_size == 0) {
-        std::fill_n(skip_decode_, batch_size, true);
+    const Tensor runtime_top_k = runtime_args->isExist("runtime_top_k") ? runtime_args->at("runtime_top_k") : Tensor();
+    const size_t runtime_top_k_size = runtime_top_k.size();
+    uint         min_top_k          = runtime_top_k_size > 0 ? runtime_top_k.min<uint>() : 0;
+    skip_all_                       = false;
+
+    if (runtime_top_p_size == 0 || min_top_k > 0) {
```
[the remark this answers is not in the document]
Comment (zhyncs) [reply src/turbomind/layers/sampling_layers/TopPSamplingLayer.cu:147]: @lzhangzz @irexyc @lvhan028 Do you have any suggestions?

[the remark this answers is not in the document]
Comment (irexyc) [reply src/turbomind/layers/sampling_layers/TopPSamplingLayer.cu:146]: The topk and topk sampling layers are fused layers. Either one of these two layers will do topk and topk sampling. 

If top_k > 0, chose topk sampling layer, otherwise chose topp sampling layer.

[the remark this answers is not in the document]
Comment (zhyncs) [reply src/turbomind/layers/sampling_layers/TopPSamplingLayer.cu:146]: ok

File: src/turbomind/layers/sampling_layers/TopKSamplingLayer.cc [hunk 6]
```diff
@@ -143,49 +142,57 @@ void TopKSamplingLayer<T>::setup(const size_t batch_size, const size_t beam_widt
...
-    float top_p = runtime_top_p_size == 0 ? 0.0f : runtime_top_p.getVal<float>();
+    // skip topk setup & forward if all top_k is zero and all top_p is not zero
+    BaseSamplingLayer<T>::setup(batch_size, beam_width, runtime_args);
+
+    if (h_runtime_top_k_.size() < batch_size) {
+        h_runtime_top_k_.resize(batch_size);
+        h_runtime_top_p_.resize(batch_size);
+    }
+
+    uint  top_k = runtime_top_k_size > 0 ? runtime_top_k.max<uint>() : 0;
```
Comment (zhyncs) [inline src/turbomind/layers/sampling_layers/TopKSamplingLayer.cc:166]: Why is the value here taken as max? I think `runtime_top_k.getVal<uint>()` is ok. Since it used at `setup_topk_runtime_args`, it only takes effect when `top_ks_size == 1`, but selecting max does not affect the correctness of the final result.

Comment (irexyc) [reply src/turbomind/layers/sampling_layers/TopKSamplingLayer.cc:166]: You are right, it should be runtime_top_k.getVal<uint>().

File: src/turbomind/layers/sampling_layers/TopKSamplingLayer.cc [hunk 7]
```diff
@@ -143,49 +142,57 @@ void TopKSamplingLayer<T>::setup(const size_t batch_size, const size_t beam_widt
...
     const Tensor runtime_top_p = runtime_args->isExist("runtime_top_p") ? runtime_args->at("runtime_top_p") : Tensor();
-    const size_t runtime_top_k_size = runtime_top_k.size();
     const size_t runtime_top_p_size = runtime_top_p.size();
+    const Tensor runtime_top_k = runtime_args->isExist("runtime_top_k") ? runtime_args->at("runtime_top_k") : Tensor();
+    const size_t runtime_top_k_size = runtime_top_k.size();
+    uint         max_top_k          = runtime_top_k_size > 0 ? runtime_top_k.max<uint>() : 0;
+    float        min_top_p          = runtime_top_p_size > 0 ? runtime_top_p.min<float>() : 0.0f;
+    skip_all_                       = false;
+
+    if (max_top_k == 0 && min_top_p != 0.0f) {
```
[the remark this answers is not in the document]
Comment (irexyc) [reply src/turbomind/layers/sampling_layers/TopKSamplingLayer.cc:153]: My mistake. Only the topk kernel is fused kernel which will do topp sampling. The topp kernel will not use topk parameter.

[the remark this answers is not in the document]
Comment (zhyncs) [reply src/turbomind/layers/sampling_layers/TopKSamplingLayer.cc:153]: ok

File: src/turbomind/models/llama/LlamaBatch.cc [hunk 8]
```diff
@@ -627,11 +627,11 @@ void LlamaBatch<T>::Initialize(GenerationState& g)
     g.partial_context_legnth = partial_len;
     g.unique_ids             = std::move(unique_ids);
     g.finished_count         = 0;
+    g.skip_init_sampling     = skip_init_sampling;
 
     if (!skip_init_sampling) {
         g.max_init_ctx_len = max_context_len;
         g.step             = max_context_len;
-        InitializeSampling(g);
```
Comment (lvhan028) [inline src/turbomind/models/llama/LlamaBatch.cc:634]: Is there any necessity that we move `intializeSampling` from `Initialize` to `Forward`?

File: src/turbomind/layers/sampling_layers/TopKSamplingLayer.cc (hunk 7 above)
Comment (lvhan028) [inline src/turbomind/layers/sampling_layers/TopKSamplingLayer.cc:153]: what does `max_top_k==0 && min_top_p==0.0f` indicate?

File: src/turbomind/layers/sampling_layers/TopKSamplingLayer.cc [hunk 9]
```diff
@@ -143,49 +142,57 @@ void TopKSamplingLayer<T>::setup(const size_t batch_size, const size_t beam_widt
...
+        skip_all_ = true;
+        return;
+    }
 
-    uint  top_k = runtime_top_k.max<uint>();
-    float top_p = runtime_top_p_size == 0 ? 0.0f : runtime_top_p.getVal<float>();
+    // skip topk setup & forward if all top_k is zero and all top_p is not zero
```
Comment (lvhan028) [inline src/turbomind/layers/sampling_layers/TopKSamplingLayer.cc:158]: Should we move this comment before `if (max_top_k == 0 && min_top_p != 0.0f)`?

File: src/turbomind/layers/sampling_layers/TopKSamplingLayer.cc (hunk 7 above)
Comment (lvhan028) [reply src/turbomind/layers/sampling_layers/TopKSamplingLayer.cc:153]: I get it in `setup_topk_runtime_args`

Comment (lvhan028) [reply src/turbomind/layers/sampling_layers/TopKSamplingLayer.cc:153]: What's the relation between `top_k` and `n_logprobs`? Is `n_logprobs` not greater than `top_k?
I just wonder if `skip_all` affects returning `logits` or `logprobs`

**Comment by lvhan028** [comment]
LGTM.
But I'd like to initiate a discussion: can we only maintain the BaseSamplingLayer? This consideration arises Since I have observed a significant amount of duplicated code between TopKSamplingLayer and TopPSamplingLayer

**Review by lvhan028 — APPROVED** [review]

**Comment by irexyc** [comment]
> But I'd like to initiate a discussion: can we only maintain the BaseSamplingLayer?

I write some suggestions in https://github.com/InternLM/lmdeploy/pull/1966

**Comment by zhyncs** [comment]
cc @lzhangzz

File: src/turbomind/layers/sampling_layers/TopKSamplingLayer.cc (hunk 9 above)
Comment (lzhangzz) [reply src/turbomind/layers/sampling_layers/TopKSamplingLayer.cc:158]: +1

**Review by lzhangzz — APPROVED** [review]

**Comment by zhyncs** [comment]
After this refactor, the whl size reduced by about 3MB. @irexyc cc @lzhangzz 

new
https://github.com/zhyncs/lmdeploy-build/releases/tag/203f9e4

old
https://github.com/zhyncs/lmdeploy-build/releases/tag/f613814
